### PR TITLE
Flutter Plugin Tool supports multi-platform plugin config

### PR DESCRIPTION
**Pull request by iskakaushik**
## Description

This addresses https://github.com/flutter/flutter/issues/10288 partially.

I'm currently re-using a Plugin pojo. There will be some follow up work
here:

1. [DONE] Migrate the current pojo to have a `Map<String, PlatformPlugin>` and have
impls for: `AndroidPlatformPlugin`, `IOSPlatformPlugin`,
      `MacOSPlatformPlugin`. They can each contain their own fields.

2. Add more meaningful error messages to the tool when we encounter any
invalid configurations. Currently there is not much error handling done.

3. [DONE] Address https://github.com/flutter/flutter/issues/38628

## Related Issues

1. https://github.com/flutter/flutter/issues/10288
2. https://github.com/flutter/flutter/issues/38628 
3. https://github.com/flutter/flutter/issues/33597

## Tests

I added the following tests: `plugin_parsing_test.dart`.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by iskakaushik** [comment]
@amirh i'm staging this PR to get any initial feedback you had about the vision.

I do not intend to land it in its current state. I will have follow-up work to address the things listed in the PR description.

File: packages/flutter_tools/lib/src/plugins.dart
Comment (iskakaushik) [inline packages/flutter_tools/lib/src/plugins.dart]: `V1` and `V2` might be confusing. Might be better to stick with just `Legacy` and `MultiPlatform` as the canonical names for these.

File: packages/flutter_tools/lib/src/plugins.dart
Comment (iskakaushik) [inline packages/flutter_tools/lib/src/plugins.dart]: might be worth creating a new exception type for failures relating to parsing a plugin. I haven't seen any custom exceptions in the `flutter_tools` project so I stuck with `Exception`.

**Comment by iskakaushik** [comment]
I'm also changing the current parameters and namespacing them as they relate to the new format:

- `androidPackage` becomes `android -> package`
- `iosPrefix` becomes `ios -> classPrefix`

**Review by jonahwilliams — COMMENTED** [review]
This approach as written seems to require only well known keys, but platforms themselves are an open set. Furthermore, we'd also like each platform to be able to require/supply its own keys, so the main manifest class should probably not be responsible for parsing them.

File: packages/flutter_tools/lib/src/plugins.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/plugins.dart]: legacy and multiplat sounds reasonable.

File: packages/flutter_tools/lib/src/platform_plugins.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/platform_plugins.dart]: This seems like a better approach for splitting up the plugins. It would also be helpful to add some sort of unknown/unspecified platform that handles keys we don't recognize.

File: packages/flutter_tools/lib/src/platform_plugins.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/platform_plugins.dart]: nit, place the fields above the first method.

File: packages/flutter_tools/lib/src/platform_plugins.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/platform_plugins.dart]: For the well known plugin types, please add documentation for the expected keys and their meaning

File: packages/flutter_tools/lib/src/plugins.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/plugins.dart]: I would restructure this so that we can grab unspecified platforms as well as known platforms.

File: packages/flutter_tools/lib/src/platform_plugins.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/platform_plugins.dart]: Please remove `classPrefix` for macOS; we don't need to support it, it was just a stand-in for indicating support. We don't have the legacy concerns of iOS.

Ideally we shouldn't have it for iOS in the new format either, since it should just be part of the `pluginClass` now that that's platform-specific, but because the template, very unfortunately, was designed such that the header filename doesn't include the class prefix, I think we're stuck with it forever (standard ObjC practice is for the filename to match the full class name, including prefix, so it's nonstandard). On the plus side, with Swift being the default template now, over time that will be less common.

File: packages/flutter_tools/lib/src/plugins.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/plugins.dart]: Remove; we should not consider macOS to be part of the legacy format. There are ~no public macOS plugins that I don't control (and those that I do have deliberately not been published formally), and I've always been clear that things on desktop are still subject to breakage without fallback. I would prefer that when this lands, only the new format supports macOS, for consistency with the rest of desktop and to minimize legacy code.

(If you can coordinate with me when you land this so I can fix FDE plugins at the same time that would be helpful, but it's not critical.)

File: packages/flutter_tools/lib/src/plugins.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/plugins.dart]: Remove, per comments in the other file.

File: packages/flutter_tools/lib/src/plugins.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/plugins.dart]: Drop this whole block and associated comment, per my comment in the other file.

File: packages/flutter_tools/lib/src/plugins.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/plugins.dart]: Remove comment, since this PR is fixing it.

File: packages/flutter_tools/test/general.shard/plugin_parsing_test.dart
Comment (stuartmorgan) [inline packages/flutter_tools/test/general.shard/plugin_parsing_test.dart]: Please add macOS coverage to the new format test.

File: packages/flutter_tools/lib/src/platform_plugins.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/platform_plugins.dart]: `classPrefix` isn't actually a required parameter (and I wouldn't expect Swift plugins to use it) so this comment is somewhat misleading as written.

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]

**Comment by iskakaushik** [comment]
@stuartmorgan , I have removed all references to macOS in legacy format, i did not realize that none of them were published. I also included macOS plugin in the test that I wrote.

@jonahwilliams , I have added the documentation about the required fields for the plugins. I have also filed https://github.com/flutter/flutter/issues/38853 to address support for unknown/unsupported plugins. I will add support for them in a follow-up PR. 

I have also resolved the other comments you have given me. PTAL and thanks!

File: packages/flutter_tools/lib/src/platform_plugins.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/platform_plugins.dart]: This should be a `YamlMap` to work right?

Comment (iskakaushik) [reply packages/flutter_tools/lib/src/platform_plugins.dart]: It also works with `dynamic yaml`. Changing it to `YamlMap yaml` does make it better. Fixed!

**Review by stuartmorgan-g — APPROVED** [review]
LGTM, but please wait for @jonahwilliams's approval as well.

I'll try to prep the FDE change that will handle the macOS breaking change tonight.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]
Sorry, a couple of things I missed that I noticed while doing the FDE change:

- [This check](https://github.com/flutter/flutter/blob/69acc0f54575765574318d48074de9a23d60af3f/packages/flutter_tools/templates/cocoapods/Podfile-macos#L44) needs to be updated. I think it's fine to just remove the comment and s/macosPrefix/macos/. Ideally we should do a full parse to avoid possible collisions, but at this point there's no other 'macos' key so I think we can keep it simple for now.
- While searching for any other places that might be referenced I found [this schema](https://github.com/flutter/flutter/blob/d0e45a2369e9b7603c5d8ca33280128fba01de38/packages/flutter_tools/schema/pubspec_yaml.json#L56), which should be updated too.

**Review by amirh — COMMENTED** [review]
LGTM overall, left a few nits.

File: packages/flutter_tools/lib/src/platform_plugins.dart
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/platform_plugins.dart]: We can't AFAICT; see [this comment](https://github.com/flutter/flutter/pull/38632#discussion_r315001504)

Comment (amirh) [reply packages/flutter_tools/lib/src/platform_plugins.dart]: Though we won't be breaking any existing code if we do it right now, we would make the migration that have to happen anyway harder (I'd think not by much).

On a related note - Stuart do you think it would be useful to have a (potentially optional) `fileName` parameter here and for macOS?

Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/platform_plugins.dart]: > Though we won't be breaking any existing code if we do it right now, we would make the migration that have to happen anyway harder (I'd think not by much).
I don't follow. If you remove the parameter, then it's impossible to describe existing ObjC plugins using the new format. Are you suggesting we require everyone to rename their source files when migrating?

> On a related note - Stuart do you think it would be useful to have a (potentially optional) `fileName` parameter here and for macOS?

It's not clear to me how replacing `classPrefix` with `fileName` helps; it seems like the net effect is that we are still stuck with an extra parameter that we would rather not have, but with the added downside that migration between old and new formats is less obvious.

I don't want it for macOS; I can't see any compelling non-legacy reason to support custom filenames for source files when the convention of filename=class name is simple, and widely used in general.

Comment (amirh) [reply packages/flutter_tools/lib/src/platform_plugins.dart]: > > Though we won't be breaking any existing code if we do it right now, we would make the migration that have to happen anyway harder (I'd think not by much).
> 
> I don't follow. If you remove the parameter, then it's impossible to describe existing ObjC plugins using the new format. Are you suggesting we require everyone to rename their source files when migrating?

Yes.

> 
> > On a related note - Stuart do you think it would be useful to have a (potentially optional) `fileName` parameter here and for macOS?
> 
> It's not clear to me how replacing `classPrefix` with `fileName` helps; it seems like the net effect is that we are still stuck with an extra parameter that we would rather not have, but with the added downside that migration between old and new formats is less obvious.
> 
> I don't want it for macOS; I can't see any compelling non-legacy reason to support custom filenames for source files when the convention of filename=class name is simple, and widely used in general.

Makes sense, I didn't have a preference here, just wanted to check because if we would want to eventually support that we should not introduce extra migration pain right now.

Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/platform_plugins.dart]: If you're okay with including a file rename step in the migration for most existing plugins, then I would agree with removing classPrefix. (The plugin template for language=objc would need to be updated accordingly.) That would need to be prominently documented somewhere.

Comment (amirh) [reply packages/flutter_tools/lib/src/platform_plugins.dart]: Let's start without `classPrefix`, when this PR makes it to stable we essentially start a "soft migration" period (as we still support the old schema). If we learn that the migration is too painful we could later add `classPrefix` (it shouldn't be breaking for plugins that already migrated without `classPrefix`).

We will definitely need documentation and announcements around this migration.

Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/platform_plugins.dart]: > Let's start without `classPrefix`

Then we'll need to change the `objc` plugin template (even though it's no longer the default, it can still be selected) to create the files with the prefixes included in the same PR that changes the `pubspec.yaml` in the template. (Which doesn't need to be this PR, but if it's not going to be, then let's get an issue tracking that so it doesn't get forgotten.)

File: packages/flutter_tools/lib/src/plugins.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/plugins.dart]: This does not return a List, it returns an Iterable

Comment (amirh) [reply packages/flutter_tools/lib/src/plugins.dart]: `.toList`? (and update the first type)

**Review by stuartmorgan-g — APPROVED** [review]

**Review by amirh — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/plugins.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/plugins.dart]: else not needed.

File: packages/flutter_tools/lib/src/plugins.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/plugins.dart]: Flip the sense and un-indent the larger block:
```dart
if (!yaml.containsKey('platform')) {
  return _validateLegacyYaml(yaml);
}
// no else, un-indent.
```

File: packages/flutter_tools/lib/src/plugins.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/plugins.dart]: else not needed.

File: packages/flutter_tools/lib/src/plugins.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/plugins.dart]: If `yaml['platforms']` is not a `YamlMap`, then this call will throw a runtime type error. Please add a check, and return an error instead.

File: packages/flutter_tools/lib/src/plugins.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/plugins.dart]: ditto. Check the type of `yaml[AndroidPlugin.kConfigKey]` etc. and add appropriate error messages.

File: packages/flutter_tools/test/general.shard/flutter_manifest_test.dart
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/flutter_manifest_test.dart]: Please add tests that cover the cases where these maps are malformed, esp as noted above where they are not maps.

**Pull request merged**
